pkg/querier/queryrange/queryable.go

```diff
@@ -19,18 +20,27 @@ const (
 type ShardedQueryable struct {
 	Req     Request
 	Handler Handler
+
+	sharededQuerier *ShardedQuerier
 }
 
 // Querier implements Queryable
 func (q *ShardedQueryable) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
-	return &ShardedQuerier{ctx, q.Req, q.Handler}, nil
+	q.sharededQuerier = &ShardedQuerier{Ctx: ctx, Req: q.Req, Handler: q.Handler, ResponseHeaders: map[string][]string{}}
+	return q.sharededQuerier, nil
+}
+
+func (q *ShardedQueryable) getResponseHeaders() map[string][]string {
+	return q.sharededQuerier.ResponseHeaders
 }
 
 // ShardedQuerier is a an implementor of the Querier interface.
 type ShardedQuerier struct {
-	Ctx     context.Context
-	Req     Request
-	Handler Handler
+	Ctx                context.Context
```

Comment:
Instead of adding the cache generation implementation into the ShardedQuerier can you create a new querier that wraps this functionality?